autolab/Autolab
# Update Ruby to 3.2.2, Misc fixes

## Summary

### Summary generated by Reviewpad on 04 Jan 24 10:51 UTC
This pull request includes the following changes:

- The file "edit.html.erb" in the "app/views/announcements" directory has changes related to the `form_for` and `render` methods.
- The "Annotation" model has changes in the "update_non_autograded_score" method affecting how "annotation_delta" is calculated.
- The file "app/views/assessments/show.html.erb" has a small diff related to the `render` line.
- The "nginx.conf" file shows a change in the Ruby version used for passenger.
- The "submissions_controller_spec.rb" file has changes in the parameters passed to the "get :index" method.
- The "metrics_controller.rb" file has changes in the "get_watchlist_instances" and "get_watchlist_configuration" methods.
- The "courses_controller.rb" file has changes in the "add_users_from_emails" and "write_cuds" methods.
- The "rails_helper.rb" file in the "spec" directory has changes related to setting up the Capybara driver for testing with Chrome.
- The "courses_api_spec.rb" file has changes related to finding "course_user_data" using the "find_by" method.
- The "course_user_datum.rb" file has changes in the "find_by" method and object initialization.
- The "new.html.erb" file has a change in the `render` statement, affecting the `locals` hash.
- The "oauth_device_flow_request.rb" file has a change in the `create_request` method.
- The "app/views/users/edit.html.erb" file has a change in the `render` method, removing the `f` variable from the `locals` hash.
- The "device_flow_requests_spec.rb" file has changes in the `device_code` parameter passed to the `get :authorize` request.
- The "courses.rb" file has changes in the `FactoryBot.create` calls, missing arguments.
- The "WatchlistInstance" model has changes related to query conditions and attribute assignments.
- The "getjob.html.erb" file has changes in the rendered code, replacing file paths with `[redacted]`.
- The "tango_mock.rb" file modifies the `open_result` hash, removing a value and adding an extra closing bracket.
- The "submissions_controller.rb" file has changes related to the `view` method, including modifications to variable assignments.
- The "app/views/devise/registrations/edit.html.erb" file has changes in the `render` method and added `readonly` attribute.
- The "edit.html.erb" file has a change in the `render` statement, affecting the `path` local variable.
- The "Gemfile" shows changes in gem versions and additions/removals.
- The "index.html.erb" file in the "app/views/doorkeeper/authorized_applications" directory has a change in the `render` method.
- The "annotated_file" method, "version_over_threshold_by" method, and "group_associated_submissions" method have changes related to query conditions and variable assignments.
- The "course_user_data_api_spec.rb" file has changes in the `find_by` method arguments.

Please review these changes for any unintended modifications or errors.


## Description


Main change: Update Ruby version of Autolab from 2.7.7 to 3.2.2

- update Capybara config so that it works with new ruby version and so that js can be enabled again on selenium test
  - removed webdrivers gem since we are now using selenium 4.11+ but added webrick since spec was failing without it
- update releaseSectionGrades redirect to go to viewGradesheet for CA's section
- add some more status text / more informative flash when instructor drops student

Course user data page for instructor viewing dropped student:
<img width="1080" alt="Screenshot 2024-01-04 at 12 01 27 AM" src="https://github.com/autolab/Autolab/assets/25730111/b27f0918-5d0e-4530-9855-51c4de2e115e">

When dropping student, see flash:

<img width="1491" alt="Screenshot 2024-01-04 at 12 03 30 AM" src="https://github.com/autolab/Autolab/assets/25730111/d5558ac2-5a19-4ac2-885b-d09f141f9792">

- redact tango key in getjob

Before:
<img width="1164" alt="Screenshot 2024-01-04 at 12 05 19 AM" src="https://github.com/autolab/Autolab/assets/25730111/b059ad6d-cfe3-405e-864d-12eb64ffc0f4">

After:
<img width="1223" alt="Screenshot 2024-01-04 at 12 04 31 AM" src="https://github.com/autolab/Autolab/assets/25730111/6aabaf1d-625e-4daa-92b3-62b7c8ee5c68">


## Motivation and Context



As part of new Autolab release, ensure that Ruby is up to date and is within security EOL.

## How Has This Been Tested?




Ran 
```
rbenv install $(cat .ruby-version)
```
then 
```
bundle install
```
to update. 

Used google spreadsheet for manual testing strategy: [doc](https://docs.google.com/spreadsheets/d/1dedvmzBFWrrRw0z9iSnj0czBbkZqZUU2ZU0F-7WYqhY/edit#gid=1108780221)

Passes all spec tests.

Dropping student:
- try dropping student as instructor, see that notice flash indicates you dropped student + course user datum page indicates student is dropped
- when undropping, no flash should appear

getjob:
- submit job to autolab, see that upon viewing job the tango key isn't exposed after courselabs in input files or job parameters

releaseSectionGrades:
- try to release section grades for already released grades as a CA, see that you get an error flash and are redirect to the gradesheet for your section

## Checklist:
